### PR TITLE
DAOS-11604 control,mgmt,bio: Add dmg functionality for VMD LED management

### PR DESCRIPTION
**Pull request by tanabarr**
New dmg functionality for VMD LED management:

$ dmg storage led identify [--reset] <DEVICES>
$ dmg storage led check <DEVICES>

--reset: Reset the status LED on a VMD SSD back to the previous state
             (OFF/normal or ON/faulty) after issuing an identify command.
DEVICES: Comma-separated list of device-UUIDs or VMD backing PCI address

led-identify will set LED state to QUICK_BLINK or reset to the previous
state. led-check will display the current state of the status LED on a
VMD backing SSD. 

LED state is now displayed in dmg storage query list-devices,
dmg storage set nvme-faulty and dmg storage nvme replace cmd output. dmg storage identify vmd command has been removed.
Configurable timeout parameter to be added in a subsequent patch, default timeout changed from one to two minutes.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>
Co-authored-by: Sydney Vanda <sydney.m.vanda@intel.com>

### CODE REVIEW

**Comment by mjmac** [comment]
@tanabarr: This is a massive PR (+6k,-2k LOC !!). Is there any way to break it into smaller chunks for more targeted reviews? I think it's going to be difficult to get a quality set of reviews as-is.

**Comment by tanabarr** [comment]
> @tanabarr: This is a massive PR (+6k,-2k LOC !!). Is there any way to break it into smaller chunks for more targeted reviews? I think it's going to be difficult to get a quality set of reviews as-is.

I will try to break down, thanks.

**Comment by tanabarr** [comment]
I have broken down the PR into commits grouping changes by language (C/Go) and type (protobuf/source/test).
```
commit 55860927d0053a686583b5727796d43fdba1f005 (HEAD -> tanabarr/daos-8002_dmg)
    DAOS-8002 control: Add dmg commands for VMD LED management - C changes
 15 files changed, 506 insertions(+), 686 deletions(-)

commit 2fdcbe312fcfcfc0edb7f5046a0330ab98f6ba90
    DAOS-8002 control: Add dmg commands for VMD LED management - Go test file changes
 11 files changed, 729 insertions(+), 488 deletions(-)

commit 6f3f005c23d998ab52c3d5ae6ed306529802d0dd
    DAOS-8002 control: Add dmg commands for VMD LED management - Go source changes
 14 files changed, 576 insertions(+), 409 deletions(-)

commit 8dd24dce6c6c7ee605d9c06c3ef34eba2af3753d
    DAOS-8002 control: Add dmg commands for VMD LED management - Auto-generated protobuf-c
 4 files changed, 3866 insertions(+), 625 deletions(-)

commit 71c064e67cd35cc98f767513d2eaed7bfe82a153
    DAOS-8002 control: Add dmg commands for VMD LED management - Auto-generated protobuf-go
 3 files changed, 666 insertions(+), 1164 deletions(-)

commit 00c4ddc94f59d185932d86ef58a6909309708995
    DAOS-8002 control: Add dmg commands for VMD LED management - Protobuf definition changes
4 files changed, 130 insertions(+), 112 deletions(-)
```

As we can see from the summary the majority of changes listed in the stats are in the auto generated protobuf codes. Two main contributing factors are:
- The addition of `smd.pb-c` header and source files to the `src/bio` directory (approved by @NiuYawei ) needed in order to centralize LED and NVMe state constant definitions in protobuf enums.
- The consolidation of SMD and NVMe-health protobuf message structures used in Go code results in a large number of deletions (and simplification of code).

The actual number of Go lines changed is `14 files changed, 576 insertions(+), 409 deletions(-)`. As reference, @mjmac 's recently landed commit contained the following number of changes:
```
commit d7a909e6abae052424ef6333efa92f7ad5e5c082
Author: Michael MacDonald <mjmac.macdonald@intel.com>
Date:   Thu Sep 22 07:57:58 2022 -0400
    DAOS-11479 control: Add dmg system exclude (#10235)
 30 files changed, 1132 insertions(+), 508 deletions(-)
```

Hopefully now the PR is an acceptable form to review, I propose that to reduce the burden @mjmac and @kjacque can review the Go code and poto changes (commits  6f3f005c23d998ab52c3d5ae6ed306529802d0dd , 2fdcbe312fcfcfc0edb7f5046a0330ab98f6ba90 , 00c4ddc94f59d185932d86ef58a6909309708995) whilst @NiuYawei and @knard-intel can review the C changes (commit 55860927d0053a686583b5727796d43fdba1f005 ). Does this sound reasonable?

File: src/include/daos_srv/control.h
Comment (tanabarr) [inline src/include/daos_srv/control.h:8]: will fix if re-pushed

File: src/include/daos_srv/control.h
Comment (tanabarr) [inline src/include/daos_srv/control.h:21]: as above

File: src/control/server/ctl_smd_rpc.go
Comment (kjacque) [inline src/control/server/ctl_smd_rpc.go:502]: Seems like a good argument for breaking the commands off into their own function at this level, too.

File: src/control/server/instance_drpc.go
Comment (kjacque) [inline src/control/server/instance_drpc.go:268]: typo - indicaes -> indicates

**Review by kjacque — CHANGES_REQUESTED** [review]
Only reviewed the Golang changes in this pass. I have a few concerns -- mostly that it seems like control.SmdQuery is becoming a switch for multiple different commands. I think these should be broken up at the control API level at minimum, and maybe at the protobuf level as well.

File: src/proto/ctl/smd.proto
Comment (tanabarr) [reply src/proto/ctl/smd.proto:91]: as far as I'm aware we expect the version of dmg and daos_server to be the same as per the current interoperability guidelines, therefore this should be safe, @mjmac is that correct?

File: src/proto/ctl/smd.proto
Comment (tanabarr) [reply src/proto/ctl/smd.proto:147]: this line should be removed

File: src/proto/ctl/smd.proto
Comment (tanabarr) [reply src/proto/ctl/smd.proto:167]: transport address is what is used throughout SPDK, PCI address is one type of transport address, others being NVMoF. I'm using the transport just because sometime in the future we are supposed to be supporting NVMoF

File: src/control/cmd/dmg/pretty/printers.go
Comment (tanabarr) [reply src/control/cmd/dmg/pretty/printers.go:42]: I wasn't one hundred percent comfortable with the solution but it enables the display to be tailored for the LED command output without having to add extra bloat regarding HostStorage structures and packing /unpacking methods.

File: src/control/cmd/dmg/pretty/storage.go
Comment (tanabarr) [reply src/control/cmd/dmg/pretty/storage.go:181]: I'm following the current model of the SMD query commands which end up calling `cmd.makeRequest()` which in turn prints out the SmdInfo. Supplying the PrintConfigOption enables us to continue using this concise form otherwise there would be some duplication creating a near identical `makeRequest` function just for the two LED commands.

File: src/control/cmd/dmg/storage_query.go
Comment (tanabarr) [reply src/control/cmd/dmg/storage_query.go:208]: this isn't new code it was just relocated, maybe we can address this separately?

File: src/control/cmd/dmg/storage_query.go
Comment (tanabarr) [reply src/control/cmd/dmg/storage_query.go:244]: this is also code that wasn't changed but relocated, maybe I can address this and the above issue to separate PR?

File: src/control/cmd/dmg/storage_query.go
Comment (tanabarr) [reply src/control/cmd/dmg/storage_query.go:274]: yes I think that's correct, I've been considering renaming it but held back because it's used in so many places, will change

File: src/control/lib/control/server_meta.go
Comment (tanabarr) [reply src/control/lib/control/server_meta.go:55]: Yes this has been a significant issue for me, I started breaking them out and then realised that we would need to duplicate a lot of code in order to have a dedicated route for the LED commands e.g. in `lib/control/server_meta.go` add `LedQueryResp`, associated `addHostResponse()`, then we will need to add a LedQuery RPC to the control service and the associated handler in `server/ctl_smd_rpc.go` (similar to SmdQuery). When all the changes unravelled it seemed like the better path was just to use the flags based on the limitations of the design.

File: src/control/server/ctl_smd_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_smd_rpc.go:258]: Again something I pondered over, SetFaulty has its own drpc handler which sets both the internal bio dev state and led state. In the handler  the parameters for the set LED call are set explicitly so we don't need to pass them in. It seemed strange for SetFaulty drpc handler to accept a LED state to set  in the request given the handler function should always be setting it to faulty.

File: src/control/server/ctl_smd_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_smd_rpc.go:502]: I think I've already tried to explain the rationale for this in previous comment replies.

File: src/control/server/instance_drpc.go
Comment (tanabarr) [reply src/control/server/instance_drpc.go:268]: done

File: src/proto/ctl/smd.proto
Comment (mjmac) [reply src/proto/ctl/smd.proto:91]: > as far as I'm aware we expect the version of dmg and daos_server to be the same as per the current interoperability guidelines, therefore this should be safe, @mjmac is that correct?

This is correct. The interop requirements for the foreseeable future dictate that we support +1/-1 for client (including the agent)/server, but otherwise all components will be the same version. Any messages that are only used between admin/server or server/engine are fine to modify without worrying about backward compatibility. We just need to be careful about modifying messages that are used by both the agent and the server.

File: src/proto/ctl/smd.proto
Comment (mjmac) [reply src/proto/ctl/smd.proto:166]: In that case, it might be better to update the comment.

Comment (tanabarr) [reply src/proto/ctl/smd.proto]: done

**Review by NiuYawei — APPROVED** [review]
The C changes look good to me.

**Review by knard-intel — COMMENTED** [review]
The overall looks good for me (for what I understand).
Just some minor comments to clarify.

File: src/control/server/ctl_smd_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_smd_rpc.go:351]: If the user specifies a UUID for a SMD device that is on a PCI address that is also specified in the input list then I just filter it out, do you think we need to return an error in this case?

File: src/control/server/ctl_smd_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_smd_rpc.go:234]: We could do but even if the check is not done here an empty list of engine instances will be returned from `FilterInstancesByRankSet()` and the function will return error on L242 printing the bogus rank #. I agree that it would be an improvement to fail earlier so I will update if I have to repush.

File: src/control/server/ctl_smd_rpc.go
Comment (knard-intel) [reply src/control/server/ctl_smd_rpc.go:351]: fair enough

File: src/control/server/ctl_smd_rpc.go
Comment (knard-intel) [reply src/control/server/ctl_smd_rpc.go:234]: fair enough

File: src/control/server/ctl_smd_rpc.go
Comment (knard-intel) [reply src/control/server/ctl_smd_rpc.go:351]: At the end, from what I understand of the usage of the smdMapDevIDsToEngine(), it seems that there is no associated actions which could be silently filtered out. If my understanding is correct, than a Debug (or Notice) message is probably more than enough.

**Review by knard-intel — APPROVED** [review]
LGFM, thanks for the explanation.

**Review by NiuYawei — APPROVED** [review]

File: src/control/server/storage/bdev.go
Comment (mjmac) [inline src/control/server/storage/bdev.go:110]: nit: Is there a reason to make this type specifically about VMD LEDs? What if we add support for SCSI/SAS/SATA LEDs? Or some other type of device? As with the previous comment, I think it would be better to keep VMD contained as an implementation detail rather than exposing it in these exported types.

**Review by mjmac — COMMENTED** [review]
As discussed during the standup, it would probably be best to use this as an opportunity to make a clean distinction between Query/Action RPCs for storage devices. I don't know if LED query and management necessarily needs to be split out; I could see it fitting into the other device query/action RPCs.

File: src/control/cmd/dmg/storage.go
Comment (tanabarr) [reply src/control/cmd/dmg/storage.go:28]: done

File: src/control/server/storage/bdev.go
Comment (tanabarr) [reply src/control/server/storage/bdev.go:110]: done

File: src/proto/ctl/smd.proto
Comment (tanabarr) [reply src/proto/ctl/smd.proto:164]: the reason I kept it separate is that it's used extensively on the C side and looks relevant names look a lot tidier without the prefix

**Review by NiuYawei — DISMISSED** [review]
The C changes look good to me.

**Review by knard-intel — APPROVED** [review]
LGTM

**Comment by tanabarr** [comment]
> LGTM. Have you run all of the new commands with the `--json` output option? We just need to make sure that works before landing.

Found an issue during manual testing so it's not ready to land yet, thanks.

**Comment by sydneyvanda** [comment]
@tanabarr Sorry just checking back in on this after leave. 

> 
> --reset: Reset the status LED on a VMD SSD back to the previous state
> (OFF/normal or ON/faulty) after issuing an identify command.

In the original PR I was suggesting the reset option as a manual reset back to OFF/NORMAL state (to avoid not having an option to reset the LEDs back to normal if something acts up or accidentally gets set in testing). I was needing to do that quite a bit at the SPDK command level when LEDs were misbehaving/left in incorrect states/test case not completed successfully/etc. Not really a typical use case, but is there still a way to do that now? 
Is this reset command just to override the automatic reset of 1 min (or other configurable time param)?

**Comment by tanabarr** [comment]
> @tanabarr Sorry just checking back in on this after leave.
> 
> > --reset: Reset the status LED on a VMD SSD back to the previous state
> > (OFF/normal or ON/faulty) after issuing an identify command.
> 
> In the original PR I was suggesting the reset option as a manual reset back to OFF/NORMAL state (to avoid not having an option to reset the LEDs back to normal if something acts up or accidentally gets set in testing). I was needing to do that quite a bit at the SPDK command level when LEDs were misbehaving/left in incorrect states/test case not completed successfully/etc. Not really a typical use case, but is there still a way to do that now? Is this reset command just to override the automatic reset of 1 min (or other configurable time param)?

The design discussion and decisions are mostly covered in https://daosio.atlassian.net/browse/DAOS-11604 . I've gone back and forth on a couple of issues and in consultation with @NiuYawei it was decided that `identify --reset` should be provided to revert the action of an identify command but there shouldn't be a production command to reset a faulty state as either the device gets replaced physically or the `replace` command can be used with old and new uuid commandline option value set the same as a workaround to revive the device. Do you think that covers the use cases sufficiently? I'm also working on a configurable commandline timeout option to be added in a subsequent PR.

**Review by mjmac — COMMENTED** [review]
It looks OK to me. Would like to get some feedback from @NiuYawei and @sydneyvanda as well as some manual verification with real hardware before landing.

**Comment by tanabarr** [comment]
> It looks OK to me. Would like to get some feedback from @NiuYawei and @sydneyvanda as well as some manual verification with real hardware before landing.

I've performed a significant amount of manual verification on wolf-156 and have requested review from Niu.

**Comment by tanabarr** [comment]
ma®'e%v1--t2pco 

> > I've performed a significant amount of manual verification on wolf-156 and have requested review from Niu.
> 
> With visual verification that the LEDs are toggled as expected? This isn't my area, so perhaps this isn't necessary because it would really be testing VMD, but it seems to me that it would be a good idea to verify that we're getting the actual physical state changes that we expect with the new code.

No, only using the SPDK API to report the LED activity is as expected. I was hoping to land the PR to master before asking those in charge of the test plan to visually verify on the hardware. SPDK is extensively used and tested and should be reliable in what it reports.

File: docs/admin/administration.md
Comment (kjacque) [inline docs/admin/administration.md:630]: minor typo: SSDs -> SSD

File: src/bio/bio_internal.h
Comment (kjacque) [inline src/bio/bio_internal.h:18]: nit - Anything from this file actually used in this header? I'd lean toward including headers only in files that need them.

File: src/mgmt/srv_drpc.c
Comment (kjacque) [inline src/mgmt/srv_drpc.c:2379]: Is there a ticket for this?

File: src/control/lib/control/server_meta.go
Comment (kjacque) [inline src/control/lib/control/server_meta.go:61]: From an API perspective, why not make this a slice? These are device UUIDs, right? Maybe UUIDs would be the best name in that case.

File: src/control/lib/control/server_meta.go
Comment (kjacque) [inline src/control/lib/control/server_meta.go:275]: suggestion - If we're only allowed to do one at a time, maybe defining opcodes would be a better model.

**Review by kjacque — APPROVED** [review]
A couple minor items and suggestions, nothing I would block on.

File: docs/admin/administration.md
Comment (tanabarr) [reply docs/admin/administration.md:630]: will correct in follow-on PR

File: src/bio/bio_internal.h
Comment (tanabarr) [reply src/bio/bio_internal.h:18]: no, the function that did was added and removed during the PR commits, I will remove in the following PR or if I have to re-push

File: src/mgmt/srv_drpc.c
Comment (tanabarr) [reply src/mgmt/srv_drpc.c:2379]: This isn't something I added, I just moved it. @NiuYawei or @sydneyvanda do you have a ticket number or this? I will update in the following PR unless I have to re-push.

File: src/control/lib/control/server_meta.go
Comment (tanabarr) [reply src/control/lib/control/server_meta.go:61]: At this stage this field is expected to carry either a single UUID for set-faulty or replace manage commands, but a list of UUIDs and/or PCI addresses for the LED variants. I can update the JSON tag from uuid->ids in the follow on PR but I think it's actually not used. The field could be changed to a slice but I don't see a strong reason to given the positional argument will be comma separated and we want to keep the unpacking and validation of tokens within the control API layer.

File: src/mgmt/srv_drpc.c
Comment (kjacque) [reply src/mgmt/srv_drpc.c:2379]: Okay, no worries.

File: src/control/lib/control/server_meta.go
Comment (kjacque) [reply src/control/lib/control/server_meta.go:61]: Fair enough.

**Review by NiuYawei — APPROVED** [review]
The C changes looks good to me.

File: src/mgmt/srv_drpc.c
Comment (sydneyvanda) [reply src/mgmt/srv_drpc.c:2379]: yes this is https://daosio.atlassian.net/browse/DAOS-6283

**Review by sydneyvanda — APPROVED** [review]
LGTM. Was not able to manually test yet on my end, but code changes and logic looks good. Can test post-landing I assume.

File: src/bio/bio_device.c
Comment (NiuYawei) [inline src/bio/bio_device.c:877]: This was for tracking the LED identify (blinking) start time, see bio_led_event_monitor(). I think that's why you observed blinking doesn't get reset anymore?

Comment (tanabarr) [reply src/bio/bio_device.c:877]: I'm addressing this in the follow-up PR for configurable timeout.

Comment (tanabarr) [reply src/bio/bio_device.c:877]: addressed in https://github.com/daos-stack/daos/pull/10768

File: docs/admin/administration.md
Comment (tanabarr) [reply docs/admin/administration.md:630]: addressing in https://github.com/daos-stack/daos/pull/10768

File: src/bio/bio_internal.h
Comment (tanabarr) [reply src/bio/bio_internal.h:18]: addressing in https://github.com/daos-stack/daos/pull/10768

File: src/mgmt/srv_drpc.c
Comment (tanabarr) [reply src/mgmt/srv_drpc.c:2379]: updated on PR https://github.com/daos-stack/daos/pull/10768

File: src/control/lib/control/server_meta.go
Comment (tanabarr) [reply src/control/lib/control/server_meta.go:61]: updated the JSON tag name in PR https://github.com/daos-stack/daos/pull/10768

**Pull request merged**
